op-deployer/pkg/deployer/verify/artifacts.go

```diff
@@ -0,0 +1,137 @@
+package verify
+
+import (
+	"encoding/json"
+	"fmt"
+	"path"
+	"strings"
+
+	"github.com/ethereum-optimism/optimism/op-chain-ops/foundry"
+)
+
+type contractArtifact struct {
+	ContractName     string
+	CompilerVersion  string
+	OptimizationUsed bool
+	OptimizationRuns int
+	EVMVersion       string
+	StandardInput    string
+	ConstructorArgs  string
+}
+
+// Map state.json struct's contract field names to forge artifact names
+var contractNameExceptions = map[string]string{
+	"OptimismPortalImpl":          "OptimismPortal2",
+	"L1StandardBridgeProxy":       "L1ChugSplashProxy",
+	"L1CrossDomainMessengerProxy": "ResolvedDelegateProxy",
+	"Opcm":                        "OPContractsManager",
+}
+
+func getArtifactName(name string) string {
+	lookupName := strings.TrimSuffix(name, "Address")
+
+	if artifactName, exists := contractNameExceptions[lookupName]; exists {
+		return artifactName
+	}
+
+	lookupName = strings.TrimSuffix(lookupName, "Proxy")
+	lookupName = strings.TrimSuffix(lookupName, "Impl")
+	lookupName = strings.TrimSuffix(lookupName, "Singleton")
+
+	// If it was a proxy and not a special case, return "Proxy"
+	if strings.HasSuffix(name, "ProxyAddress") {
+		return "Proxy"
+	}
+
+	return lookupName
+}
+
+func (v *Verifier) getContractArtifact(name string) (*contractArtifact, error) {
+	artifactName := getArtifactName(name)
+	artifactPath := path.Join(artifactName+".sol", artifactName+".json")
+
+	v.log.Info("Opening artifact", "path", artifactPath, "name", name)
+	f, err := v.artifactsFS.Open(artifactPath)
+	if err != nil {
+		return nil, fmt.Errorf("failed to open artifact: %w", err)
+	}
+	defer f.Close()
+
+	var art foundry.Artifact
+	if err := json.NewDecoder(f).Decode(&art); err != nil {
+		return nil, fmt.Errorf("failed to decode artifact: %w", err)
+	}
+
+	// Add all sources (main contract and dependencies)
+	sources := make(map[string]map[string]string)
+	for sourcePath, sourceInfo := range art.Metadata.Sources {
+		remappedKey := art.SearchRemappings(sourcePath)
+		sources[remappedKey] = map[string]string{"content": sourceInfo.Content}
+		v.log.Debug("added source contract", "originalPath", sourcePath, "remappedKey", remappedKey)
+	}
+
+	var optimizer struct {
+		Enabled bool `json:"enabled"`
+		Runs    int  `json:"runs"`
+	}
+	if err := json.Unmarshal(art.Metadata.Settings.Optimizer, &optimizer); err != nil {
+		return nil, fmt.Errorf("failed to parse optimizer settings: %w", err)
+	}
+
+	standardInput := map[string]interface{}{
```

Comment:
Can we use real structs rather than `map[string]interface{}`?

Reply:
Updated in https://github.com/ethereum-optimism/optimism/pull/14359/commits/fda4d60b72fe3a30c354d28b8524cee700077a30